File: 실버 2 (5문제)/1260_DFS와 BFS.py
# 참고 링크 : https://moon-gd.tistory.com/60
import queue
import sys

q_input = sys.stdin.readline
myQueue = queue.Queue()

vertex, edge, start = map(int, q_input().split())

# 연결 정보 저장할 배열 생성
connect = []

# 방문 배열 생성
dfs_visited = []
bfs_visited = []

for _ in range(vertex + 1):
    connect.append([])
    dfs_visited.append(False)
    bfs_visited.append(False)


# 연결 정보 저장하기
for _ in range(edge):
    v1, v2 = map(int, q_input().split())
    connect[v1].append(v2)
    connect[v2].append(v1)


for i in range(1, vertex + 1):
    connect[i].sort()


def dfs(point):
    dfs_visited[point] = True
    print(point, end=' ')

    for i in connect[point]:
        if not dfs_visited[i]:
            dfs(i)


# dfs 실행
dfs(start)

# 개행 출력
print()

# BFS
myQueue.put(start)
bfs_visited[start] = True


while myQueue.qsize():
    v = myQueue.get()
    print(v, end=' ')

    for i in connect[v]:
        if bfs_visited[i]:
            continue

        myQueue.put(i)
        bfs_visited[i] = True

File: 실버 2 (6문제)/1260_DFS와 BFS.c
// 참고 링크 : https://moon-gd.tistory.com/60
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#define MAX_VERTEX 1001
#define MAX_EDGE 10000

int connect[MAX_VERTEX][MAX_VERTEX];
int visited[MAX_VERTEX];
int stack[MAX_EDGE];
int queue[MAX_EDGE];

int vertex, edge, start, top = 0;
int front = 0, rear = 0;

void dfs(int);

int main() {

	// 입력 받기
	scanf("%d %d %d", &vertex, &edge, &start);
	int v1, v2, i;
	for (i = 0; i < edge; i++) {	
		scanf("%d %d", &v1, &v2);

		// 연결 정보 저장
		connect[v1][v2] = 1;
		connect[v2][v1] = 1;
	}

	// dfs
	// 재귀 활용
	dfs(start);
	putchar('\n');

	// bfs
	// queue 활용
	// 이 때부터는 방문 배열의 의미가 반대로 쓰임
	// 0 : 방문,  1 : 방문 X
	
	// queue에 시작점 삽입 + 시작점 방문 표시
	queue[front++] = start;
	visited[start] = 0;

	while (front > rear) {
		v2 = queue[rear++];

		printf("%d ", v2);
		
		for (i = 1; i <= vertex; i++) {
			// 연결되어있는 정점이며, 방문되지 않은 경우
			if (connect[v2][i] && visited[i]) {
				// 큐 삽입 + 방문 표시
				queue[front++] = i;
				visited[i] = 0;
			}
		}
	}

	return 0;
}

void dfs(int v) {
	// 방문 표시
	visited[v] = 1;

	// 출력
	printf("%d ", v);

	for (int i = 1; i <= vertex; i++) {
		if (connect[v][i] && visited[i] == 0) {
			dfs(i);
		}
	}
}